From 3c12518f20b5d20c9d9f20fb43275a02ab0db8b1 Mon Sep 17 00:00:00 2001
From: Axtel Sturnclaw <sturnclaw@protonmail.com>
Date: Thu, 6 Aug 2026 15:27:58 -0400
Subject: [PATCH] Patch: add ImGuiWindowFlags_NoCaptureMouse

- Allows mouse inputs to pass through hovered windows to the host application.
- Windows using this should be at the BOTTOM of the window stack if they are transparent.
- This is ultimately a sledgehammer solution.
---
 contrib/imgui/imgui.cpp | 6 ++++--
 contrib/imgui/imgui.h   | 1 +
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/contrib/imgui/imgui.cpp b/contrib/imgui/imgui.cpp
index 09971db0e..db6ec0035 100644
--- a/contrib/imgui/imgui.cpp
+++ b/contrib/imgui/imgui.cpp
@@ -5664,7 +5664,9 @@ void ImGui::UpdateHoveredWindowAndCaptureFlags(const ImVec2& mouse_pos)
     }
     else
     {
-        io.WantCaptureMouse = (mouse_avail && (g.HoveredWindow != NULL || mouse_any_down)) || has_open_popup;
+        const bool hovered_skip_capture = (g.HoveredWindow != NULL && (g.HoveredWindow->Flags & ImGuiWindowFlags_NoCaptureMouse));
+        const bool hovered_skip_clicked = hovered_skip_capture && (g.ActiveId == 0 || g.ActiveId == g.HoveredWindow->MoveId);
+        io.WantCaptureMouse = (mouse_avail && (g.HoveredWindow != NULL && !hovered_skip_capture || mouse_any_down && !hovered_skip_clicked)) || has_open_popup;
         io.WantCaptureMouseUnlessPopupClose = (mouse_avail_unless_popup_close && (g.HoveredWindow != NULL || mouse_any_down)) || has_open_modal;
     }
 
@@ -8664,7 +8666,7 @@ bool ImGui::Begin(const char* name, bool* p_open, ImGuiWindowFlags flags)
         // Set default BgClickFlags
         // This is set at the end of this function, so UpdateWindowManualResize()/ClampWindowPos() may use last-frame value if overridden by user code.
         // FIXME: The general intent is that we will later expose config options to default to enable scrolling + select scrolling mouse button.
-        window->BgClickFlags = (flags & ImGuiWindowFlags_ChildWindow) ? parent_window->BgClickFlags : (g.IO.ConfigWindowsMoveFromTitleBarOnly ? ImGuiWindowBgClickFlags_None : ImGuiWindowBgClickFlags_Move);
+        window->BgClickFlags = (flags & ImGuiWindowFlags_ChildWindow) ? parent_window->BgClickFlags : (g.IO.ConfigWindowsMoveFromTitleBarOnly || flags & ImGuiWindowFlags_NoCaptureMouse ? ImGuiWindowBgClickFlags_None : ImGuiWindowBgClickFlags_Move);
 
         // We fill last item data based on Title Bar/Tab, in order for IsItemHovered() and IsItemActive() to be usable after Begin().
         // This is useful to allow creating context menus on title bar only, etc.
diff --git a/contrib/imgui/imgui.h b/contrib/imgui/imgui.h
index bc598a38d..529a374b7 100644
--- a/contrib/imgui/imgui.h
+++ b/contrib/imgui/imgui.h
@@ -1232,6 +1232,7 @@ enum ImGuiWindowFlags_
     ImGuiWindowFlags_NoNavFocus             = 1 << 17,  // No focusing toward this window with keyboard/gamepad navigation (e.g. skipped by Ctrl+Tab)
     ImGuiWindowFlags_UnsavedDocument        = 1 << 18,  // Display a dot next to the title. When used in a tab/docking context, tab is selected when clicking the X + closure is not assumed (will wait for user to stop submitting the tab). Otherwise closure is assumed when pressing the X, so if you keep submitting the tab may reappear at end of tab bar.
     ImGuiWindowFlags_NoDocking              = 1 << 19,  // Disable docking of this window
+    ImGuiWindowFlags_NoCaptureMouse         = 1 << 20,  // [PIONEER] Disable hovering this window capturing the mouse for ImGui's use.
     ImGuiWindowFlags_NoNav                  = ImGuiWindowFlags_NoNavInputs | ImGuiWindowFlags_NoNavFocus,
     ImGuiWindowFlags_NoDecoration           = ImGuiWindowFlags_NoTitleBar | ImGuiWindowFlags_NoResize | ImGuiWindowFlags_NoScrollbar | ImGuiWindowFlags_NoCollapse,
     ImGuiWindowFlags_NoInputs               = ImGuiWindowFlags_NoMouseInputs | ImGuiWindowFlags_NoNavInputs | ImGuiWindowFlags_NoNavFocus,
-- 
2.54.0

